platform-hub-api/app/services/kubernetes/token_claim_service.rb

```diff
@@ -0,0 +1,75 @@
+module Kubernetes
+  module TokenClaimService
+    extend TokenLookup
+    extend self
+
+    module Errors
+      class TokenNotFound < StandardError; end
+      class TokenAlreadyDefinedForCluster < StandardError; end
+      class TokenInvalid < StandardError; end
+      class TokenAlreadyClaimed < StandardError; end
+    end
+
+    def claim_token(user, token)
+      claim = lookup(token)
+
+      if claim.nil?
+        raise Errors::TokenNotFound, "Token `#{token}` not found!"
+      end
+
+      if claim.user.nil?
+        begin
+          migrate_claimed_token_to_user_kubernetes_identity(user, claim)
+          remove_claimed_token_from_static_list(claim)
+        rescue => e
+          raise
+        end
+        [ claim.cluster, "Claimed `#{claim.cluster}` token." ]
+
+      elsif claim.user.present?
+        raise Errors::TokenAlreadyClaimed, "Token already claimed!"
+      end
+    end
+
+    private
+
+    def migrate_claimed_token_to_user_kubernetes_identity(user, claim)
+      identity = 
+        user.identity(:kubernetes) || 
+        user.identities.create!(
+          provider: :kubernetes,
+          external_id: user.email,
+          data: {tokens: []}
+        )
+
+      if identity.data['tokens'].find {|t| t['cluster'] == claim.cluster}
+        raise Errors::TokenAlreadyDefinedForCluster, 
+              "User `#{user.email}` already have a token assigned for `#{claim.cluster}` cluster!"
```

Comment:
Maybe we should overwrite the token? There's a possibility (albeit small) where:
- user requests new token
- admin uses the hub to grant them a new token for that cluster
- user finds their old token, and tries to claim it

… maybe this isn't an important use case to cover? (The only negative impact I can think of is we end up with stale token in the static list, which we can always clean up later).